compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/PropertyReferenceLowering.kt

```diff
@@ -0,0 +1,347 @@
+import org.jetbrains.kotlin.backend.common.ir.copyTo
+import org.jetbrains.kotlin.backend.common.ir.createImplicitParameterDeclarationWithWrappedDescriptor
+import org.jetbrains.kotlin.backend.common.lower.createIrBuilder
+import org.jetbrains.kotlin.backend.common.phaser.makeIrFilePhase
+import org.jetbrains.kotlin.backend.jvm.JvmBackendContext
+import org.jetbrains.kotlin.descriptors.*
+import org.jetbrains.kotlin.ir.builders.*
+import org.jetbrains.kotlin.ir.builders.declarations.*
+import org.jetbrains.kotlin.ir.declarations.*
+import org.jetbrains.kotlin.ir.expressions.*
+import org.jetbrains.kotlin.ir.expressions.impl.*
+import org.jetbrains.kotlin.ir.symbols.*
+import org.jetbrains.kotlin.ir.types.*
+import org.jetbrains.kotlin.ir.types.impl.*
+import org.jetbrains.kotlin.ir.util.*
+import org.jetbrains.kotlin.ir.visitors.*
+import org.jetbrains.kotlin.load.java.JvmAbi
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.types.Variance
+
+internal val propertyReferencePhase = makeIrFilePhase(
+    ::PropertyReferenceLowering,
+    name = "PropertyReference",
+    description = "Construct KProperty instances returned by expressions such as A::x and A()::x"
+)
+
+internal class PropertyReferenceLowering(val context: JvmBackendContext) : ClassLoweringPass {
+    // TODO: join IrLocalDelegatedPropertyReference and IrPropertyReference via the class hierarchy?
+    private val IrMemberAccessExpression.getter: IrFunctionSymbol?
+        get() = (this as? IrPropertyReference)?.getter ?: (this as? IrLocalDelegatedPropertyReference)?.getter
+
+    private val IrMemberAccessExpression.setter: IrFunctionSymbol?
+        get() = (this as? IrPropertyReference)?.setter ?: (this as? IrLocalDelegatedPropertyReference)?.setter
+
+    private val IrMemberAccessExpression.field: IrFieldSymbol?
+        get() = (this as? IrPropertyReference)?.field
+
+    private val IrMemberAccessExpression.signature: String
+        // Plain Java fields do not have a getter, but can be referenced nonetheless.
+        get() = getter?.owner?.let { context.state.typeMapper.mapSignatureSkipGeneric(it.descriptor).toString() }
+            ?: TODO("plain Java field signature")
+
+    private fun IrFunction.visibleFrom(scope: Scope): Boolean =
+        // Local declarations (delegated property accessors in this case) have LOCAL visibility; so long as the property itself
+        // is visible, then so is the accessor.
+        visibility == Visibilities.LOCAL || visibility.isVisible(Visibilities.ALWAYS_SUITABLE_RECEIVER, descriptor, scope.scopeOwner)
+
+    private val plainJavaClass =
+        context.getIrClass(FqName("java.lang.Class")).owner
+
+    private val reflectionClass =
+        context.getIrClass(FqName("kotlin.jvm.internal.Reflection")).owner
+
+    private val getOrCreateKotlinClass =
+        reflectionClass.declarations.find { it.descriptor.name == Name.identifier("getOrCreateKotlinClass") }
```

Comment:
Please use `reflectionClass.functions` here and below, and remove casts in `buildReflectedContainerReference`